### DISCUSSION TITLE
Library Request: molstar a protein viewer.

### DISCUSSION THREAD

**Discussion by zachcp** [observablehq/framework#1061] [Q&A] [answered]
Hi Team Observable,

- I'd like to be able to embed [molstar](https://molstar.org/) viewers. 
- According to the [stubborn importer troubleshooter](https://observablehq.com/@observablehq/module-require-debugger) this is a tricky import.
- theres are some nice precedents for other static site embeds - e.g. [quarto-molstar](https://github.com/jmbuhr/quarto-molstar)

I'll work on some import options but it might make for a nice addition to the library section.

thanks!
zach cp

**Comment by mbostock** [comment, marked as answer]
Here’s how I got it working. First I downloaded the molstar library, putting the files in `docs`:

```sh
curl -o docs/molstar.js https://molstar.org/viewer/molstar.js
curl -o docs/molstar.css https://molstar.org/viewer/molstar.css
```

Then I created the following page following the [plugin documentation](https://molstar.org/docs/plugin/):

````md
<link rel="stylesheet" type="text/css" href="molstar.css">

```js
import "./molstar.js";

const molstar = window.molstar;
```

<div id="app" style="position: relative; width: 800px; height: 600px;"></div>

```js echo
molstar.Viewer.create("app", {
  layoutIsExpanded: false,
  layoutShowControls: false,
  layoutShowRemoteState: false,
  layoutShowSequence: true,
  layoutShowLog: false,
  layoutShowLeftPanel: true,
  viewportShowExpand: true,
  viewportShowSelectionMode: false,
  viewportShowAnimation: false,
  pdbProvider: "rcsb",
  emdbProvider: "rcsb",
}).then(viewer => {
  viewer.loadPdb("7bv2");
  viewer.loadEmdb("EMD-30210", {detail: 6});
});
```
````

It looks like this:

<img width="1121" alt="Screenshot 2024-03-14 at 6 23 11 PM" src="https://github.com/observablehq/framework/assets/230541/c1ce107e-d020-4b25-a95a-0c0de954a540">

You currently need to use `import` instead of a `script` tag, but see #1035. (And `import` is probably better anyway, since it’s async.)

**Comment by zachcp** [comment]
wow. you are..... amazing. fast and amazing.

**Comment by zachcp** [comment]
HI Good Morning @mbostock ,

I ran into an issue extending the above example to use a local file. For example if we add a local PDB and attempt to use it we get encounter an issue related to file not found. Can you point me to how to let Observable know about local static assets in this case?

thank you,
zach cp

```sh
curl -o docs/7BV2.pdb  https://files.rcsb.org/download/7BV2.pdb
```


```js
molstar.Viewer.create("app", {
  layoutIsExpanded: false,
}).then((viewer) => {
    viewer.loadStructureFromUrl("7BV2.pdb", "pdb");   // <--  file is not found
});
```


![image](https://github.com/observablehq/framework/assets/1260667/2d40d4b4-738f-4103-8954-275ba81e828d)

**Reply by zachcp** [reply]
```js
       viewer.loadStructureFromUrl("_file/7BV2.pdb", "pdb");    // <---- okay. giving the direct filepath in the compiled site will work...
```

**Reply by Fil** [reply]
I think this will only work in preview; to make sure that the file is recognized in the build, you will have to register it in a way or another. For example by doing something like:

`async (viewer) => viewer.loadStructureFromUrl(await FileAttachment("7BV2.pdb").url(), "pdb");`

(sorry I can't test rn, so this might not work)

**Reply by zachcp** [reply]
Thank you. Will try and report back.

**Reply by zachcp** [reply]
- Apparantly my understanding of async javascript is terrible.
- when I switch to the code above I get no results and also no console error. 
- i'm trying to sprinkle `then()` here and there to force the resolution.

**Reply by Fil** [reply]
works for me… I'll share the code in a minute

![Capture d’écran 2024-03-15 à 16 05 30](https://github.com/observablehq/framework/assets/7001/2a0ae08e-6968-4e0f-bc58-09fdf6431d57)

**Reply by zachcp** [reply]
got it as well!  TY

**Reply by zachcp** [reply]
I had the entire async FN call internal  to the `then` block. Durf.

**Reply by zachcp** [reply]
```js echo
molstar.Viewer.create("app", {
  layoutIsExpanded: false,
  layoutShowControls: false,
  layoutShowRemoteState: false,
  layoutShowSequence: true,
  layoutShowLog: false,
  layoutShowLeftPanel: true,
  viewportShowExpand: true,
  viewportShowSelectionMode: false,
  viewportShowAnimation: false,
  pdbProvider: "rcsb",
  emdbProvider: "rcsb",
}).then(async (viewer) => {
        viewer.loadStructureFromUrl(await FileAttachment("7BV2.pdb").url(), "pdb")  
});

**Reply by Fil** [reply]
the pdb file is not enough to host locally, since this is also loading
`https://maps.rcsb.org/em/emd-30210/cell?detail=6`
and 
`https://files.wwpdb.org/pub/emdb/structures/EMD-30210/header/emd-30210.xml`

**Reply by zachcp** [reply]
( theres a bunch of details that I can work out once I have a working example.  This package does a LOT and was designed to retrieve files from Web APIs. I'm pretty sure theres knobs to turn those all off. )

**Reply by zachcp** [reply]
Would you like me to add an example to the documentation page?

**Reply by Fil** [reply]
I feel that this library is too advanced/niche for something that we'd want to have in the main documentation. But it's nice to have it documented here and elsewhere (hint: your blog?) to help future users find it.

**Reply by Fil** [reply]
I've updated the solution and added it to pangea (my big list of examples): https://observablehq.observablehq.cloud/pangea/party/molstar

**Reply by mbostock** [reply]
@Fil I’m seeing some errors in that example. You can’t use an HTML code block for this:

```html
<link rel="stylesheet" type="text/css" href="npm:molstar/build/viewer/molstar.css">
```

An HTML code block is rendered client-side (because HTML code blocks are dynamic), so the npm: URL isn’t resolved, and then you see this error in the console:

```
Failed to load resource: net::ERR_UNKNOWN_URL_SCHEME npm:molstar/build/viewer/molstar.css:1
```

I then see a separate error here:

```js
FileAttachment("molstar/7bv2.pdb").href
```

I think the path to the file attachment is wrong. It should be:

```js
FileAttachment("7bv2.pdb").href
```

**Comment by Fil** [comment]
Thanks for spotting the CSS error, this was my mistake: a last-minute change so I could display the code, and I did not verify. The missing file error was due to #1367; observable cloud deploy is case sensitive — the code was requesting `7Bv2.pdb` but the file is named `7BV2.pdb`.

**Reply by Fil** [reply]
I'm still seeing an error, which looks internal to molstar and does not seem to prevent it from working.

```
Error: Unexpected token. Expected data_, loop_, or data name.
    at t.f (molstar.49ad7967.js:2:2512771)
```

**Reply by zachcp** [reply]
You guys are fantastic.
